[82,22] S3 download/upload support for xlsx,xlsm and geojson

j-gillam opened the pull request:
# S3 download/upload support for xlsx,xlsm and geojson

This PR closes the issues: 82 and 22. 

### Description

S3 download/upload support for xlsx,xlsm and geojson.
Added:

- Updates to S3.py file for uploading/downloading xlsx,xlsm and geojson.
- Updated setup with new package requirements. 
- Added tests for uploading and downloading the files.
- Added dummy xlsx,xlsm and geojson files. 

### Checklist:

- [x] I have followed [Contributor Guidelines](https://github.com/nestauk/nesta_ds_utils/blob/dev/CONTRIBUTING.md)
- [x] I have updated the documentation to include any new functionality I have added or modified
- [x] I have written unit tests for any new functionality I have added
- [x] I have ran and passed all tests locally with `>> pytest`
- [x] I have checked that all tests ran successfully on the Github after I pushed

j-gillam committed:
[82,22] S3 download/upload support for xlsx,xlsm and geojson

j-gillam committed:
[82,22] S3 download/upload support for xlsx,xlsm and geojson

sqr00t commented on setup.cfg:
```diff
@@ -33,6 +36,7 @@ nlp =
 test =
   pytest==7.1.3
   moto[s3]==4.0.7
+  shapely==2.0.2
```
```suggestion
```
this is already included by %(s3)s as it is added in line 24

j-gillam replied:
Woops missed that! Thanks!

sqr00t commented on setup.cfg:
```diff
@@ -20,6 +20,9 @@ install_requires =
 [options.extras_require]
 s3 =
   boto3==1.24.93
+  openpyxl==3.0.9
+  shapely==2.0.2
+  geopandas==0.13.2
```
It may be best to add a `gis` environment to include `shapely` and `geopandas`
I'd suggest adding `openpyxl` to another environment `io_extras` to reduce bloat, or add to install_requires
Another possibility is adding [all] to pandas in install_requires i.e. `pandas[all]==1.5.1`
We could also follow [how pandas adds](https://github.com/pandas-dev/pandas/blob/main/pyproject.toml) each engine/ extra and enable it downstream in this package

```suggestion
gis =
  geopandas==0.13.2
io_extras = 
  openpyxl==3.0.9
```

j-gillam replied:
Could we just add pandas[excel]==1.5.1 ? And actually thinking about it, geopandas has shapely as a dependency, so can just remove it right?

sqr00t replied:
Yep!

emily-bicks commented on nesta_ds_utils/loading_saving/S3.py:
```diff
@@ -46,9 +47,15 @@ def _df_to_fileobj(df_data: pd.DataFrame, path_to: str, **kwargs) -> io.BytesIO:
     else:
         raise Exception(
-            "Uploading dataframe currently supported only for 'csv' and 'parquet'."
+            "Uploading dataframe currently supported only for 'csv', 'parquet', 'xlsx', xlsm' and 'geojson'."
```
may be worth specifying this as a separate function given it's a geopandas df rather than a pandas df?

j-gillam replied:
I don't mind either way! I think a geopandas df still counts as a pandas df. The only difference being an extra geometry column that has a particular typing and that the package itself has extra functionality for geojsons etc. But happy to separate into another function, if that makes readability better?

j-gillam replied:
I think maybe? On the geojson question, I've never needed it but someone might! So will add it in!

sqr00t replied:
Latest geopandas docs indicates that geopandas df is a separate object type even though the library is supposedly built on top of pandas.

I think you'd need to add an instance type check condition for `gpd.GeoDataFrame` for each `fnmatch` branch, and raise an appropriate error message whether the filetype is implemented for GeoDataFrames. Or just handle `GeoDataFrame` types above the `fnmatch` branches (or move the handling to another function separately, then calling your new functions conditionally).

something like:
```
if isinstance(df_data, gpd.GeoDataFrame):
    your_new_geo_functions(params)
```

and also, importing `Union` from `typing`, and modify the function parameters type hints to:
`def _df_to_fileobj(df_data: Union[pd.DataFrame, gpd.GeoDataFrame], ...`

Linking below some reference to supported IO operations for GeoPandas:
- [User Guide](https://geopandas.org/en/stable/docs/user_guide/io.html#)
- [IO methods](https://geopandas.org/en/stable/docs/reference/io.html)
- [GeoPandas.GeoDataFrame instance methods](https://geopandas.org/en/stable/docs/reference/geodataframe.html#)
- Note, there is some overlap between the IO methods and "instance methods" sections, but there's a few extra `to_something` methods in the latter.

j-gillam replied:
Ah fab! So the isinstance(geodf, pd.DataFrame ) returns true for geopandas dataframe, but it doesnt work in the other direction. So it must count as pandas somewhere in it. I'll just make a separate function!

sqr00t replied:
I'm slightly confused, did you do instance checking on a gpd.GeoDataFrame or pd.DataFrame?

j-gillam replied:
Sorry! I think I've confused things woops. I had currently just checked for pd.DataFrame because it was neater and geodataframes are inherently pandas dataframes (at least according to the isinstance), but I will separate them out as its more robust and has caused confusion ahah.

sqr00t replied:
Just checked the geopandas source code as well:

You could also try:
```
from geopandas.base import GeoPandasBase
```

```
issubclass(gpd.GeoDataFrame, GeoPandasBase)
```

Since GeoDataFrame [inherits from](https://github.com/geopandas/geopandas/blob/main/geopandas/geodataframe.py#L72) GeoPandasBase and pandas DataFrame

j-gillam replied:
Ooo I like that! Thanks so much for checking everything :)

emily-bicks reviewed (approved):
Looks great to me! My only comment is do we want to make the "download_as" argument now be specific as pandas dataframe vs geopandas dataframe? And is there any reason why a user would want to return the entire geojson as a dictionary rather than just loading the features attribute into a geopandas df?

j-gillam committed:
Update setup.cfg

Removing shapely from test.

Co-authored-by: Solomon Yu <797346+sqr00t@users.noreply.github.com>

j-gillam committed:
adding updates based on comments, support for geojson to dictionary

j-gillam commented:
Hello both! I have updated the files based on your comments :) 

Things to note:

- @sqr00t pandas 1.5.1 doesn't allow extras, so I went back to your initial idea! 
- Separated the upload/download for geodataframes into separate functions. 
- Noticed a few spelling errors, so fixed those! 
- Removed `from xmlrpc.client import Boolean` for S3.py as I couldn't see where it was used? 

For adding geojson upload/download from dictionaries, I couldn't find a way to check typing for geojson specifically. As from my understanding of the [docs](https://datatracker.ietf.org/doc/rfc7946/), it is a json with specific formatting. So what I did was add a check to see if it has a "type" member and that member is one of the accepted types based on the specification in section 3. If anyone knows of a better way I am happy to update!

sqr00t committed:
chore: remove unused imports from file_ops, use bool instead of xmlrpc Boolean

sqr00t committed:
refactor: improve error types raised

sqr00t committed:
refactor: improve error types raised

sqr00t committed:
refactor: refine geodataframe instance type check

sqr00t reviewed (approved):
I've added some smaller changes:
- Removed `fnmatch` and `xmlrpc` imports from `file_ops` (likely a prior refactor artefact, great spot @j-gillam)
- Specified error types raised
- Refined instance checking to gpd.base.GeoPandasBase
![Screenshot 2023-10-31 at 11 09 38](https://github.com/nestauk/nesta_ds_utils/assets/797346/dc91f2fd-1873-4b6d-af45-867b6944b46c)

sqr00t commented on nesta_ds_utils/loading_saving/S3.py:
```diff
@@ -294,64 +365,85 @@ def download_obj(
-                "for 'csv' and 'parquet'."
+                "for 'csv','parquet','xlsx' and 'xlsm'."
+            )
+    elif download_as == "geodataframe":
```
```suggestion
    elif download_as.lower() in ["geodataframe", "gdf", "geodf", "geo_df"]:
```
geodataframe may be lengthly for end users

j-gillam replied:
I think its probably simpler to just have one option? We could go with geodf?

sqr00t commented on nesta_ds_utils/loading_saving/S3.py:
```diff
@@ -294,64 +365,85 @@ def download_obj(
                 "Download as numpy array currently supported only "
                 "for 'csv' and 'parquet'."
             )
     elif not download_as:
```
```suggestion
    if not download_as:
```
move this to the top statement

j-gillam committed:
simplifying geodataframe call and moved None condition to top

j-gillam committed:
simplifying geodataframe call and moved None condition to top

j-gillam committed:
simplifying geodataframe call and moved None condition to top
